File: leetcode-algorithms/076. Minimum Window Substring/Solution.cpp
// Given a string S and a string T, find the minimum window in S which will contain all the characters in T in complexity O(n).
//
// Example:
//
// Input: S = "ADOBECODEBANC", T = "ABC"
// Output: "BANC"
// Note:
//
// If there is no such window in S that covers all characters in T, return the empty string "".
// If there is such window, you are guaranteed that there will always be only one unique minimum window in S.
class Solution {
public:
    string minWindow(string s, string t) {
        if (s.size() < t.size()) return "";

        map<char, int> tmap;
        for (auto v : t) tmap[v]++;

        int count = 0;
        int start = 0;
        int minlen = INT_MAX;
        int pos = 0;
        for (int i = 0; i < s.size(); i++) {
            if (tmap.count(s[i])) {
                //count++;
                tmap[s[i]]--;
                if (tmap[s[i]] >= 0) count++;
            }

            while (count == t.size()) {
                int len = i - start + 1;
                if (len < minlen) {
                    minlen = len;
                    pos = start;
                }
                if (tmap.count(s[start])) {
                    tmap[s[start]]++;
                    if (tmap[s[start]] > 0) count--;
                }
                start++;
            }
        }

        return minlen == INT_MAX ? "" : s.substr(pos, minlen);
    }
};

File: leetcode-algorithms/076. Minimum Window Substring/Solution.java
class Solution {
    public String minWindow(String s, String t) {
        if (s.length() < t.length()) return "";
        Map<Character, Integer> fmap = new HashMap<>();
        for (char c : t.toCharArray()) {
            fmap.put(c, fmap.getOrDefault(c, 0) + 1);
        }

        int i = 0;
        int count = 0;

        int minlen = Integer.MAX_VALUE;
        int pos = 0;
        int start = 0;

        while (i < s.length()) {
            char c = s.charAt(i);

            if (fmap.containsKey(c)) {
                fmap.put(c, fmap.get(c) - 1);
                if (fmap.get(c) >= 0) count++;
            }

            while (count == t.length()) {
                int local = i - start + 1;
                if (local < minlen) {
                    minlen = local;
                    pos = start;
                }

                //start++;
                char v = s.charAt(start++);
                if (fmap.containsKey(v)) {
                    fmap.put(v, fmap.get(v) + 1);
                    if (fmap.get(v) > 0) count--;
                }
            }
            i++;
        }

        return minlen == Integer.MAX_VALUE ? "" : s.substring(pos, pos + minlen);
    }
}
